### ISSUE TITLE
`balena ssh` logs error, but exits 0

### ISSUE THREAD

**Issue by xginn8** [balena-io/balena-cli#1535]
- **balena CLI version:** 11.20.1
---
When a device appears as online (but is unresponsive to SSH), the CLI returns an error but does not exit uncleanly:

```
❯ balena ssh UUID ; echo $?
error: host error:
Connection to ssh.balena-devices.com closed.
0
```

**Comment by srlowe** [comment]
@xginn8 Do you know of a way to reproduce this device state for testing?

**Comment by srlowe** [comment]
@xginn8 Ah I've managed to reproduce.

**Comment by pdcastro** [comment]
@srlowe, FYI, it's possible that the balena-proxy backend is somehow getting in the way. Note the difference between using a local IP address, and a device UUID in the following examples, both for the same RPi 4 device, where the `exit` shell command is used to produce a non-zero exit code:
```
$ echo 'cat /etc/issue; exit 3;' | balena ssh 192.168.0.37 ; echo $?
balenaOS 2.48.0 \n \l

Warning: ssh process exited with non-zero code "3"
3
```
```
$ echo 'cat /etc/issue; exit 3;' | balena ssh dcbff32 ; echo $?
=============================================================
    Welcome to balenaOS
=============================================================
balenaOS 2.48.0 \n \l

0
```
If so, the following balena-proxy issues may reveal some hints at the relevant backend code:
* https://github.com/balena-io/balena-proxy/issues/465
* https://github.com/balena-io/balena-proxy/issues/466

**Comment by srlowe** [comment]
@pdcastro Yeah the backend masking something crossed my mind too.  However the local ssh also doesn't use `spawnSshAndExitOnError`, so the problem could be there too.  Will require some in depth investigation I think.

**Comment by pdcastro** [comment]
This issue can be reproduced with plain ssh, removing the balena CLI from the equation altogether:

```sh
$ ssh gh_paulo_castro@ssh.balena-devices.com host 2eb94bd6ea9a3b9b4c0442aebf7bdb18; echo $?
error: host error:
0
```

I got the above by purposely turning off `openvpn` on a balenaOS device (`systemctl stop openvpn` on a host OS prompt).

I believe this is an instance of the more generic issue where the balenaCloud proxy does not pass through the exit error code of remote command execution:

```sh
$ echo 'exit 1;' | balena ssh 2eb94bd6ea9a3b9b4c0442aebf7bdb18; echo $?
=============================================================
    Welcome to balenaOS
=============================================================
0
```

or:

```sh
$ echo 'exit 1;' | ssh gh_paulo_castro@ssh.balena-devices.com host 2eb94bd6ea9a3b9b4c0442aebf7bdb18; echo $?
=============================================================
    Welcome to balenaOS
=============================================================
0
```

Expected '1', got '0'. When `balena-proxy` is not involved, the balena CLI exits with error code '1' as expected:

```sh
$ echo 'exit 1;' | ssh -p 22222 root@192.168.1.19; echo $?
Pseudo-terminal will not be allocated because stdin is not a terminal.
1
```

So I believe that this is a backend `balena-proxy` issue, as reported in balena-io/balena-proxy/issues/661. That repo is private though, so I'd keep this issue open for visibility.

**Comment by pdcastro** [comment]
> I believe this is an instance of the more generic issue where the balenaCloud proxy does not pass through the exit error code of remote command execution.

In this particular sense, a workaround of sorts is to make the remote command print the desired exit code to stdout (in a way that does not spoil other printed and parsed output, if possible), then parse it and enforce it on the client side. For example, the following incorrectly prints '0':

```
$ echo 'false; exit;' | balena ssh 2eb94bd6ea9a3b9b4c0442aebf7bdb18; echo $?
=============================================================
    Welcome to balenaOS
=============================================================
0
```

And the workaround would be:

```
$ echo 'false; echo TESTEXITCODE=$?; exit;' | balena ssh 2eb94bd6ea9a3b9b4c0442aebf7bdb18 | sed -ne 's/^TESTEXITCODE=//p' | xargs -IX sh -c 'exit X'; echo $?
1
```

Breaking it down:

* `echo TESTEXITCODE=$?`, on the remote side, prints the correct exit code to stdout as `TESTEXITCODE=1`.
* `sed -ne 's/^TESTEXITCODE=//p'`, on the client side, parses that output and extracts value `1`, which is then piped to `xargs`, also on the client side.
* `xargs` that then executes `exit 1` in a subprocess through `sh -c`. The `-IX` flag means _"replace X with the value read from stdin"._
